Problem (CRO-578 WIP): Wallet keep loading and decrypt global state

yihuang opened the pull request:
Solution:
- No need to encrypt global state
- Store `lite::TrustedState` in global state
- Cache global state in memory when syncing.

Make some less controversial changes first.

devashishdxt commented on client-core/src/handler/default_block_handler.rs:
```diff
@@ -93,13 +91,7 @@ where
             }
         }
-
-        self.global_state_service.set_global_state(
```
I don't think this should be removed from here. (Reasons are mentioned in a different comment)

yihuang commented:
> Mostly looks fine. Only one change to update `global_state` inside `DefaultBlockHandler`.
> 
> On an overall design level, I think a better approach to solve `CRO-578` is to implement something like `CachedSledStorage` which implements both `Storage` and `SecureStorage` and internally uses something like [lru_time_cache](https://crates.io/crates/lru_time_cache) to cache previously queried values and also updates the cache when any value is updated. After implementing this, we can use `CachedSledStorage` wherever we need caching of values (even if the values are encrypted in storage, we can cache decrypted values).
> 
> Any thoughts?

I think the first thing to do is separate different levels of confidentiality.
- global state: no encryption
- public keys: encrypted with some key, can be cached in memory when syncing
- private keys: encrypted with another key, not cached in memory by default, can be cached in memory with explicit api call like `unlock_wallet`, unlock can have expire time set like ethereum api.

devashishdxt commented:
> I think the first thing to do is separate different levels of confidentiality.
> 
> * global state: no encryption
> * public keys: encrypted with some key, can be cached in memory when syncing
> * private keys: encrypted with another key, not cached in memory by default, can be cached in memory with explicit api call like `unlock_wallet`, unlock can have expire time set like ethereum api.

I agree. But I think we shouldn't tackle all the problems at once. The current task (`CRO-578`) is for repeatedly accessing and decrypting information stored in database. The things that you mentioned are important but they should not affect our decision of having a caching layer on top of storage.

yihuang commented on client-core/src/synchronizer/manual_synchronizer.rs:
```diff
@@ -195,7 +196,9 @@ where
 
-            self.global_state_service.save_trust_state(&trust_state)?;
+            global_state.trusted_state = trusted_state;
+            self.global_state_service
```
Because `trusted_state` is updated once for the whole batch, if we save other parts of the global state per block, it would not be consistent.
I think the chance of crash during the processing of a block batch which already downloaded from the network is quite low.

devashishdxt replied:
Ok.

tomtau commented:
bors r+

yihuang commented:
> > I think the first thing to do is separate different levels of confidentiality.
> > 
> > * global state: no encryption
> > * public keys: encrypted with some key, can be cached in memory when syncing
> > * private keys: encrypted with another key, not cached in memory by default, can be cached in memory with explicit api call like `unlock_wallet`, unlock can have expire time set like ethereum api.
> 
> I agree. But I think we shouldn't tackle all the problems at once. The current task (`CRO-578`) is for repeatedly accessing and decrypting information stored in database. The things that you mentioned are important but they should not affect our decision of having a caching layer on top of storage.

It is relevant, for example, it's not necessary to have a LRU cache for the public keys, we should keep them cached during the whole process of syncing.
And the lock/unlock api should only be used for private key related stuff.

Maybe the next viable step would be to just cache the current wallet and passphrase in memory during the whole syncing process, hopefully use a different passphrase for private keys later on.

tomtau commented:
the integration test fails:
```
  9 failing
6140	
6141	
6142	  1) Wallet Auto-sync
6143	       "before all" hook for "can auto-sync unlocked wallets":
6144	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6145	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6146	      at process._tickCallback (internal/process/next_tick.js:68:7)
6147	
6148	  2) Wallet transaction
6149	       "before all" hook in "Wallet transaction":
6150	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6151	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6152	      at process._tickCallback (internal/process/next_tick.js:68:7)
6153	
6154	
6155	  3) Wallet Auto-sync
6156	       "before all" hook for "can auto-sync unlocked wallets":
6157	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6158	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6159	      at process._tickCallback (internal/process/next_tick.js:68:7)
6160	
6161	  4) Wallet management
6162	       "before all" hook for "can restore hd-wallet with specified name":
6163	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6164	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6165	      at process._tickCallback (internal/process/next_tick.js:68:7)
6166	
6167	  5) Wallet transaction
6168	       "before all" hook in "Wallet transaction":
6169	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6170	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6171	      at process._tickCallback (internal/process/next_tick.js:68:7)
6172	
6173	
6174	  6) Staking
6175	       "before all" hook for "unbond of same amount and nonce from different account should have different txid":
6176	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6177	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6178	      at process._tickCallback (internal/process/next_tick.js:68:7)
6179	
6180	
6181	  7) Wallet Auto-sync
6182	       "before all" hook for "can auto-sync unlocked wallets":
6183	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6184	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6185	      at process._tickCallback (internal/process/next_tick.js:68:7)
6186	
6187	
6188	  8) Wallet management
6189	       "before all" hook for "cannot access un-existing wallet":
6190	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6191	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6192	      at process._tickCallback (internal/process/next_tick.js:68:7)
6193	
6194	  9) Wallet transaction
6195	       "before all" hook in "Wallet transaction":
6196	     Error: Error when synchronizing Default wallet: Verify error: block verify failed: InsufficientVotingPower, origin: None
6197	      at Object.exports.asyncMiddleman (test/core/utils.ts:81:9)
6198	      at process._tickCallback (internal/process/next_tick.js:68:7)
6199	
```

yihuang commented:
Fixed integration tests, which is caused by wrong genesis global state.
Also removed the misleading `Default` instance of `GlobalState`, and add a explicit `genesis` constructor to it.

tomtau commented:
bors try

tomtau commented:
looks ok, the failure in sgx-cargo-1804-hw2 is random :/

tomtau commented:
bors r+

tomtau commented:
bors retry

tomtau commented:
bors r+
